### PR TITLE
Provide a fake grpclb balancer server that can be used for testing

### PR DESCRIPTION
**Pull request by apolcyn**
context: I'm working on a grpclb-in-DNS interop test suite that uses fake a local DNS servers, backends, and balancers, in https://github.com/grpc/grpc/pull/16727, in order to test grpclb edge cases with respect to DNS, various credentials types, and different client languages. For this, I've found it convenient to write the fake balancer server in go.

This PR is mostly aimed providing something that can be used by https://github.com/grpc/grpc/pull/16727, but I'm thinking it might be useful to have around for manual testing anyways.

### CODE REVIEW

File: interop/fake_grpclb/fake_grpclb.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		grpclog.Errorf("Error receiving LoadBalanceRequest: %v", err)
+		return err
+	}
+	grpclog.Info("LoadBalancerRequest received.")
+	initialReq := lbReq.GetInitialRequest()
+	if initialReq == nil {
+		grpclog.Info("Expected first request to be an InitialRequest. Got: %v", lbReq)
+		return status.Error(codes.Code(2), "First request not an InitialRequest")
+	}
+	// gRPC clients targetting foo.bar.com:443 can sometimes include the ":443" suffix in
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: I remember this was discussed, and is expected.

If we want to remove the port, how about https://golang.org/pkg/net/#SplitHostPort?
Or we require the port number in `serviceName`.

File: interop/fake_grpclb/fake_grpclb.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		cleanedName = cleanedName[0 : len(cleanedName)-len(":443")]
+	}
+	if cleanedName != *serviceName {
+		grpclog.Info("Expected requested service name: %v. Got: %v", *serviceName, initialReq.Name)
+		return status.Error(codes.Code(2), "Bad requested service name")
+	}
+	if err := stream.Send(&lbpb.LoadBalanceResponse{
+		LoadBalanceResponseType: &lbpb.LoadBalanceResponse_InitialResponse{
+			InitialResponse: &lbpb.InitialLoadBalanceResponse{
+				ClientStatsReportInterval: &durationpb.Duration{
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: Or leave this field `nil`?

File: interop/fake_grpclb/fake_grpclb.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+					Nanos:   0,
+				},
+			},
+		},
+	}); err != nil {
+		grpclog.Errorf("Error sending initial LB response: %v", err)
+		return status.Error(codes.Code(2), "Error sending initial response")
+	}
+	serverList := make([]*lbpb.Server, len(l.BackendAddrs))
+	for i := range l.BackendAddrs {
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: If we use proto message in the server struct, this won't be necessary.

File: interop/fake_grpclb/fake_grpclb.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	grpclog.Info("Send LoadBalanceResponse: %v", res)
+	if err := stream.Send(res); err != nil {
+		grpclog.Errorf("Error sending LB response: %v", err)
+		return status.Error(codes.Code(2), "Error sending response")
+	}
+	if *shortStream {
+		return nil
+	}
+	for {
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: Do we need this now?

File: interop/fake_grpclb/fake_grpclb.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	"google.golang.org/grpc/credentials/alts"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/testdata"
+)
+
+var (
+	port        = flag.Int("port", 10000, "Port to listen on.")
+	backendIps  = flag.String("backend_addrs", "", "Comma separated list of backend IP/port addresses.")
+	useALTS     = flag.Bool("use_alts", false, "Listen on ALTS credentials.")
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: Merge this and TLS flag into one that takes a string?
Otherwise we will need to deal with `*alts==true && *tls==true`, and it's hard to add another one (insecure maybe), just like in interop client.

File: interop/fake_grpclb/fake_grpclb.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+			grpclog.Fatalf("Failed to generate credentials %v", err)
+		}
+		opts = append(opts, grpc.Creds(creds))
+	} else if *useALTS {
+		altsOpts := alts.DefaultServerOptions()
+		altsTC := alts.NewServerCreds(altsOpts)
+		opts = append(opts, grpc.Creds(altsTC))
+	}
+	rawBackendAddrs := strings.Split(*backendIps, ",")
+	if rawBackendAddrs[0] == "" {
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: What does this mean? Is this a special case or just an error?

If this is an error, it's not necessary because the following parsing would report this as well.

File: interop/fake_grpclb/fake_grpclb.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+		rawBackendAddrs = make([]string, 0)
+	}
+	parsedBackendAddrs := make([]backendAddr, len(rawBackendAddrs))
+	for i := range rawBackendAddrs {
+		ipAndPort := strings.Split(rawBackendAddrs[i], ":")
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: Use https://golang.org/pkg/net/#SplitHostPort

File: interop/fake_grpclb/fake_grpclb.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+		}
+		grpclog.Infof("Adding backend ip: %v, port: %d", ip.String(), port)
+		parsedBackendAddrs[i] = backendAddr{
+			Ip:   ip,
+			Port: int32(port),
+		}
+	}
+	server := grpc.NewServer(opts...)
+	grpclog.Infof("Begin listening on %d.", *port)
+	lis, _ := net.Listen("tcp", ":"+strconv.Itoa(*port))
```
Comment (menghanl) [inline interop/fake_grpclb/fake_grpclb.go]: Check the error here.

File: interop/fake_grpclb/fake_grpclb.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+	var err error
+	if lbReq, err = stream.Recv(); err != nil {
+		grpclog.Errorf("Error receiving LoadBalanceRequest: %v", err)
+		return err
+	}
+	grpclog.Info("LoadBalancerRequest received.")
+	initialReq := lbReq.GetInitialRequest()
+	if initialReq == nil {
+		grpclog.Info("Expected first request to be an InitialRequest. Got: %v", lbReq)
+		return status.Error(codes.Code(2), "First request not an InitialRequest")
```
[the remark this answers is not in the document]
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: yep, and done

File: interop/fake_grpclb/fake_grpclb.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Runs a fake grpclb balancer server, can be useful for testing.
```
[the remark this answers is not in the document]
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: done

File: interop/fake_grpclb/fake_grpclb.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+	serviceName = flag.String("service_name", "UNSET", "Name of the service being load balanced for.")
+)
+
+type backendAddr struct {
+	Ip   []byte
+	Port int32
+}
+
+type loadBalancerServer struct {
+	BackendAddrs []backendAddr
```
[the remark this answers is not in the document]
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: no good reason. And good point on using the proto, done

File: interop/fake_grpclb/fake_grpclb.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+	port        = flag.Int("port", 10000, "Port to listen on.")
+	backendIps  = flag.String("backend_addrs", "", "Comma separated list of backend IP/port addresses.")
+	useALTS     = flag.Bool("use_alts", false, "Listen on ALTS credentials.")
+	useTLS      = flag.Bool("use_tls", false, "Listen on TLS credentials, using a test certificate.")
+	shortStream = flag.Bool("short_stream", false, "End the balancer stream immediately after sending the first server list.")
+	serviceName = flag.String("service_name", "UNSET", "Name of the service being load balanced for.")
+)
+
+type backendAddr struct {
+	Ip   []byte
```
[the remark this answers is not in the document]
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: no good reason, removed per above comment

File: interop/fake_grpclb/fake_grpclb.go (hunk 1 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: let's chat offline. For now, I'm using SplitHostPort

File: interop/fake_grpclb/fake_grpclb.go (hunk 2 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: done

File: interop/fake_grpclb/fake_grpclb.go (hunk 3 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: good point, done

File: interop/fake_grpclb/fake_grpclb.go (hunk 4 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: The test suite that's using it needs it. But I could also leave it out now and then add it in a follow up.
It's a useful test scenario for exercising some timing edge cases, and seems to reveal some unexpected behaviors so far.

File: interop/fake_grpclb/fake_grpclb.go (hunk 5 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: hmm, I agree this is a little weird, but the current flags do allow specifying insecure (by keeping both of these flags off), and it's convenient to be able to pass the same flags to this server as for the interop server

What do you think?

File: interop/fake_grpclb/fake_grpclb.go (hunk 6 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: this was a messy way to handle the case when `--backend_addrs` is an empty string (which is actually valid and used for some test scenarios). Revised to clean this up

File: interop/fake_grpclb/fake_grpclb.go (hunk 7 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: done

File: interop/fake_grpclb/fake_grpclb.go (hunk 8 above)
Comment (apolcyn) [reply interop/fake_grpclb/fake_grpclb.go]: done
